File: packages/w2v/reference/src/training/skip_gram.c
#include "atomic_float.h"
#include "training.h"

#include <string.h>

static void copy_input(
    const Model *model,
    size_t token,
    real *destination)
{
    size_t dimension = model->embedding_dimension;
    const _Atomic uint32_t *input_row =
        model->input_embeddings + token * dimension;

    for (size_t coordinate = 0; coordinate < dimension; coordinate++)
    {
        destination[coordinate] = atomic_float_load(&input_row[coordinate]);
    }
}

static void update_input(
    Model *model,
    size_t token,
    const real *gradient)
{
    size_t dimension = model->embedding_dimension;
    _Atomic uint32_t *input_row =
        model->input_embeddings + token * dimension;

    for (size_t coordinate = 0; coordinate < dimension; coordinate++)
    {
        atomic_float_add(&input_row[coordinate], gradient[coordinate]);
    }
}

void skip_gram_train(ModelStep *step)
{
    Model *model = step->trainer->model;
    Worker *worker = step->worker;
    size_t dimension = model->embedding_dimension;
    size_t radius = context_radius(step);
    size_t center_token = step->target_token;

    for (size_t offset = 0; offset <= radius * 2; offset++)
    {
        size_t position;
        if (context_position(worker, radius, offset, &position))
        {
            size_t context_token = worker->sentence[position];
            ModelStep context_step = *step;
            context_step.target_token = context_token;

            copy_input(model, center_token, worker->hidden);
            memset(
                worker->hidden_gradient,
                0,
                dimension * sizeof(*worker->hidden_gradient));
            objective_train(
                &context_step,
                worker->hidden,
                worker->hidden_gradient);
            update_input(model, center_token, worker->hidden_gradient);
        }
    }
}

File: packages/w2v/src/training/skip_gram.rs
use super::{
    ModelStep, ObjectiveLoss, ObjectiveScratch, context_position, context_radius_with_policy,
    hierarchical_softmax, negative_sampling, shared_add,
};
use crate::{
    atomic_float,
    config::{ObjectiveKind, Status},
};

#[inline]
pub fn train(step: &mut ModelStep<'_, '_>) -> Result<ObjectiveLoss, Status> {
    let model = &step.trainer.model;
    let dimension = model.embedding_dimension;
    let radius = context_radius_with_policy(
        step.window_rng,
        step.trainer.config.window_radius,
        step.trainer.config.context_policy,
    );
    let center_token = step.target_token;
    let start = center_token * dimension;
    let input_row = &model.input_embeddings[start..start + dimension];

    let mut loss = ObjectiveLoss::default();
    for offset in 0..=radius * 2 {
        if let Some(position) =
            context_position(step.sentence.len(), step.sentence_position, radius, offset)
        {
            let context_token = step.sentence[position];
            for (coordinate, value) in input_row.iter().enumerate() {
                step.hidden[coordinate] = atomic_float::load(value);
            }
            step.hidden_gradient.fill(0.0);
            loss.add(match step.trainer.config.objective_kind {
                ObjectiveKind::HierarchicalSoftmax => hierarchical_softmax::train_with_scratch(
                    step.trainer,
                    context_token,
                    step.learning_rate,
                    step.hidden,
                    ObjectiveScratch {
                        hidden_gradient: step.hidden_gradient,
                        output_snapshot: step.output_snapshot,
                    },
                    step.observe_objective,
                ),
                ObjectiveKind::NegativeSampling => negative_sampling::train_with_scratch(
                    step.trainer,
                    context_token,
                    step.learning_rate,
                    step.negative_rng,
                    step.hidden,
                    ObjectiveScratch {
                        hidden_gradient: step.hidden_gradient,
                        output_snapshot: step.output_snapshot,
                    },
                    step.observe_objective,
                ),
            }?);
            for (coordinate, value) in input_row.iter().enumerate() {
                shared_add(
                    value,
                    step.hidden_gradient[coordinate],
                    step.trainer.config.update_strategy,
                );
            }
        }
    }
    Ok(loss)
}
